FIXED JUnit Test Cases！！Upgrade the logging, JDBC Connection Pool , SOLR+Lucene, Basic Library...etc

Yuanuo opened the pull request:
When LOCK is placed in the constructor, a second call to getInstance () will throw an error after LOCK is released or repeat initialization.
It is usually the responsibility of getInstance to initialize m_instance only once, so LOCK should be used by it.

aKandzior commented:
Thank you very much for sending this pull request.

We are currently checking how to integrate your changes. We want to keep version 10.5 compatible with Java 7, so merging against branch_10_5_x is not an option. 

We are in the process of merging against branch_11_0_x. This will take a couple of weeks to complete.

tHerrmann commented:
Thanks again for your work. We will definitely use some of your changes within our next major release.
For future pull request, please make sure the test cases run without any failures. To do this, just run the gradle task 'test' and check the results generated.
In this case, some test failures where related to the changed log4j library. A specific log4j2 configuration and some changes to our custom test log appender could resolve those. But there are many test failures within the org.opencms.search packages, that are related to the lucene update. Those were not easily resolved and point to remaining issues with this library update.

Yuanuo commented:
Sorry, I do not know "gradle test", I will try to solve the problem of not passing this test.

By the way, is the Lucene implementation under the org.opencms.search package sure it is useful? I used the process, it seems that only used Solr's implementation.

Yuanuo commented:
very very sorry,

The code was submitted without running the JUnit test case after the last upgrade.

Now that I've made some fixes and basically solved most of the problems, I ran the org.opencms.test.AllTests test in Eclipse and the result is as follows:
Runs: 896/896, Errors: 2, Failures: 3.

2 Errors are:
1) Tests for package org.opencms.i18n
org.opencms.i18n.TestCmsCoreMessageBundles
testLocale_DE_MessagesBundles (org.opencms.i18n.TestCmsCoreMessageBundles)
2) org.opencms.i18n.TestCmsModuleMessageBundles
testLocale_DE_MessagesBundles (org.opencms.i18n.TestCmsModuleMessageBundles)

These two errors may not work because many messages.properties do not have a localized version of German. So the test does not pass.

3 Failures are:
1) Tests for package org.opencms.search
org.opencms.search.TestCmsSearchInDocuments
testSearchBoostInMeta (org.opencms.search.TestCmsSearchInDocuments)
junit.framework.ComparisonFailure: expected: <... efault / search / test1. [xls]> but was: <... efault / search / test1. [doc]>
2) testSearchBoost (org.opencms.search.TestCmsSearchInDocuments)
junit.framework.ComparisonFailure: expected: <... efault / search / test1. [doc]> but was: <... efault / search / test1. [xls]>
3) Tests for package org.opencms.search.solr
org.opencms.search.solr.TestSolrSearch
testDocumentBoost (org.opencms.search.solr.TestSolrSearch)
junit.framework.AssertionFailedError: 9searchNew.txt with priority 'max' should be the first

These three issues are mainly due to the Boost feature when the index was removed from the previous version of Lucene / Solr. For the time being did not complete the implementation of the Boost feature replacement.
I'm very sorry that I did not know the Boost feature until I upgraded the code, so I did not use this feature in my project. But maybe others use it, but I did not rewrite the code after upgrading.


If possible, you just need to fix this part of Boost so that the entire JUnit test case can be tested.

Yuanuo commented:
In addition, I am really puzzled, why I submitted the Pull Request all automatically merged into this one Pull request inside. .


By the way, this code may run wrong directly "gradle test", "gradle bindist" is no problem.
The error "gradle test" is because there is a gwt-dev-2.7.0.jar in the classpath that contains many older classes, such as the Apache HttpClient class, but the HttpClient used in this version of Solrj 7.2 is Compared to the new version, so there will be NoSuchMethodError.

So when I was testing, I did not use "gradle test" but ran the JUnit test directly in Eclipse. Of course, I did not use "Gradle Nature", but "Eclipse" project with "Maven Nature".

tHerrmann commented:
Hi,
I have issue with the SOLR configuration. The referenced class org.ansj.solr.AnsjTokenizerFactory is missing from the class path. And I can't find the JAR it is located in.

Yuanuo commented:
HI.
Looks like it's a direct copy of the Solr configuration used in my project when fixing Junit Test.
That Ansj is used to do Chinese word segmentation, so did not submit.
Please use the file in this attachment to replace "test / data / WEB-INF / solr / configsets / default / conf / managed-schema" on it.

[managed-schema.zip](https://github.com/alkacon/opencms-core/files/1716343/managed-schema.zip)

Yuanuo commented:
This GitHub, probably I do not know the correct usage, all my code commits it listed on this page. I even deleted the original Fork project, Fork again, or so. Sorry for the inconvenience.

tHerrmann commented:
Thanks, the changed managed-schema file works. But there is still an issue with the new JSON implementation you introduced. org.opencms.search.solr.TestSolrFieldConfiguration.testDependencies fails because of issues reading a JSON array. Is there a specific reason you replaced the previous implementation?

Yuanuo commented:
Sorry, the preceding code did not submit complete.

This Pull Request made me very depressed. I deleted the project, restarted Fork again, and then re-modified the code, and made a Commit, it seems still not every Pull Request per Commit, so now I give up.

JSON implementation replacement, I recently used JSON for another project, and then saw some performance comparisons that looked like org.json did not perform as well as the Jackson JSON implementation. In this OpenCms, I actually did not study JSON performance issues, did not do the actual test, but felt that there was a simple way (with a small change of code) that I could replace / upgrade, and that's it.
After the code was submitted, I found some bugs, and did repair and Commit, but without Push, I deleted the original Fork project.

Today I made a lot of changes and did a lot of Gradle tests and passed them (including JSON-related Tests, except that the 3 tests except Solr Boost still fail).

By the way, I submitted a gwt-dev.jar, delete some of the Class related to apache-commons in the jar, now you can complete the test directly with "gradle test".

Yuanuo commented:
Another big change I make today is to separate the VfsDriver interface(readContent,writeContent,etc..) about the byte[]Content of CmsFile data store so that it can store some large files outside the main database, such as local files, Or other server (via FTP), database (MongoDB, etc.).

This is done because of my project, my site needs to provide the user to download some audio, video, these stored in the database may not make sense, directly stored in the local file, provided to the user to download may be better.

Did the test, JUnit Test passed.

In addition, there may be too many changes I made that make it difficult to match your company's plans. Looks like there are still many people using old versions of OpenCms in the mailing list.
However, I changed my actual needs, if some changes can be integrated into the OpenCms inside, we may have some help, so I will be very happy.

tHerrmann commented:
Hi,
we have integrated most of your changes into the new branch_11_0_x. Those are mostly the library updates including lucene/solr. We skipped some added features.
Thanks again for your work.

Yuanuo commented:
You're welcome, my pleasure.
